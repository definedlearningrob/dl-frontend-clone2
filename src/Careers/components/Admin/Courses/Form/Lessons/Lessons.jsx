import PropTypes from 'prop-types';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useField } from 'formik';

import { AvailableLessonsList } from '@dc/components/Admin/Courses/Form/Lessons/AvailableLessonsList/AvailableLessonsList';
import LessonDetailsModal from '@dc/components/Admin/Courses/Form/Lessons/Modal/Modal';
import { PaginationBar } from '@dc/components/Admin/Shared/List/PaginationBar/PaginationBar';
import SelectedLessonsList from '@dc/components/Admin/Courses/Form/Lessons/SelectedLessonsList/SelectedLessonsList';
import { LESSON_TYPES } from '@dc/resources/constants';
import { ReactComponent as LessonsIcon } from '@dc/svg/table_of_content.svg';
import { assignSteps } from '@dc/utils/assignSteps';

import SharedPaginatedLoader from '@shared/components/PaginatedLoader/PaginatedLoader';
import Card from '@shared/components/Card/Card';
import { IconContainer } from '@shared/components/IconContainer/IconContainer';
import { ListWrapper } from '@shared/components/SelectableList/ListWrapper/ListWrapper';
import { Select } from '@shared/components/Select';
import { ErrorMessage } from '@shared/components/ErrorMessage/ErrorMessage';

AdminCoursesFormLessons.propTypes = {
  clearFilter: PropTypes.func,
  fieldName: PropTypes.string,
  label: PropTypes.string,
  lessonTypeOptions: PropTypes.array,
  pagingProps: PropTypes.object,
  SearchBar: PropTypes.func,
  selectedLessonType: PropTypes.object,
  selectLessonType: PropTypes.func,
};

function AdminCoursesFormLessons({
  clearFilter,
  fieldName,
  label,
  lessonTypeOptions,
  pagingProps,
  SearchBar,
  selectedLessonType,
  selectLessonType,
}) {
  const [activeLesson, setActiveLesson] = useState(null);
  const { t } = useTranslation();
  const [lessonsInput, lessonsMeta, lessonsHelpers] = useField(fieldName);
  const isDetailsModalOpen = activeLesson !== null;

  const closeDetailsModal = () => setActiveLesson(null);
  const openItemDetails = (lesson) => setActiveLesson(lesson);
  const sortedByStep = lessonsInput.value.slice().sort((a, b) => a.step - b.step);

  const moveSurveyToBottom = (newLessons) => {
    const indexOfSurveyLesson = newLessons.findIndex(
      ({ type }) => type.toUpperCase() === LESSON_TYPES.CAREER_REVIEW_SURVEY
    );
    newLessons.push(newLessons.splice(indexOfSurveyLesson, 1)[0]);

    return assignSteps(newLessons);
  };

  const addLesson = (lesson) => {
    const { value } = lessonsInput;
    lessonsHelpers.setValue(moveSurveyToBottom([...value, lesson]));
  };

  const removeLesson = (lesson) => {
    const newLessons = lessonsInput.value.filter(({ id }) => id !== lesson.id);
    lessonsHelpers.setValue(assignSteps(newLessons));
  };

  const handleEdit = (lesson) => {
    window.open(`/admin/lesson/${lesson.id}/edit`, '_blank', 'noreferrer');
  };

  return (
    <Card>
      <Card.Header>
        <div className='flex items-center gap-xs'>
          <IconContainer
            Icon={LessonsIcon}
            className='bg-neutral-200 rounded-xs'
            paddingSize='xs'
          />
          <h4 className='m-0'>{label || t('admin.courses.lessons.label')}</h4>
        </div>
      </Card.Header>
      <div className='flex gap-sm'>
        <ListWrapper
          actions={
            <div className='flex items-center gap-xs'>
              <SearchBar
                className='!w-[20px]'
                field='name'
                placeholder={t('common.placeholders.searchBy', {
                  field: t('common.fields.common.name').toLowerCase(),
                })}
              />
              {lessonTypeOptions && (
                <Select
                  className='w-[150px]'
                  options={lessonTypeOptions}
                  showError={false}
                  value={selectedLessonType}
                  onChange={selectLessonType(pagingProps.selectPage, clearFilter)}
                />
              )}
            </div>
          }
          title={t('admin.courses.lessons.allLessons')}>
          <div className='min-h-0 flex-1'>
            <SharedPaginatedLoader.Content {...pagingProps}>
              {({ lessons }) => (
                <AvailableLessonsList
                  fieldName={fieldName}
                  lessons={lessons.nodes}
                  onAdd={addLesson}
                  onDetailsOpen={openItemDetails}
                  onEditClick={handleEdit}
                />
              )}
            </SharedPaginatedLoader.Content>
          </div>
          <PaginationBar pagingProps={pagingProps} />
        </ListWrapper>

        <ListWrapper title={`${t('common.statuses.selected')} (${sortedByStep.length})`}>
          <SelectedLessonsList
            fieldName={fieldName}
            lessons={sortedByStep}
            onChange={removeLesson}
            onDetailsOpen={openItemDetails}
            onEditClick={handleEdit}
          />
        </ListWrapper>
        {isDetailsModalOpen && (
          <LessonDetailsModal
            isOpen={isDetailsModalOpen}
            lessonId={activeLesson?.id}
            onClose={closeDetailsModal}
          />
        )}
      </div>
      {lessonsMeta.error && (
        <div className='flex justify-end mt-sm'>
          <ErrorMessage errorMessage={lessonsMeta.error} />
        </div>
      )}
    </Card>
  );
}

export default AdminCoursesFormLessons;
