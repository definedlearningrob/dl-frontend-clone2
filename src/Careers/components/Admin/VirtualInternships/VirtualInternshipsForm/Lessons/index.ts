export * from './CalendarLessons';
export * from './ExperienceOpportunityLessons';
export * from './PostExperienceLessons';
export * from './ReadinessSkillsLessons';
