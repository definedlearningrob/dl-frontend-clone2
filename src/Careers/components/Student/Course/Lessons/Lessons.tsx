import { useTranslation } from 'react-i18next';
import { useParams } from 'react-router-dom';

import { LessonCard } from '@dc/shared/LessonCard';
import { TCourseLesson } from '@dc/graphql/student/queries/course';
import { TPublicCourseLesson } from '@dc/graphql/public/queries/course';

import useQueryParams from '@shared/hooks/useQueryParams';

type Props = {
  courseProgress?: { submitted: number; total: number };
  lessons: TCourseLesson[] | TPublicCourseLesson[];
};

export const StudentCourseLessons = ({ courseProgress, lessons }: Props) => {
  const { t } = useTranslation();
  const sortedLessons = lessons.slice().sort((a, b) => a.step - b.step);
  const { id: courseId, shareId } = useParams<{ id: string; shareId: string }>();
  const {
    params: { code },
  } = useQueryParams<{ code: string }>();

  return (
    <section className='course-lessons'>
      <h3>{t('course.lessons.heading')}</h3>
      <div className='course-lessons__lessons'>
        {sortedLessons.map((lesson, index) => {
          const isSurveyPerformed =
            lesson.careerReviewSurvey !== null &&
            'performed' in lesson.careerReviewSurvey &&
            lesson.careerReviewSurvey.performed;
          const isProgress = 'progress' in lesson;

          if (courseProgress) {
            return (
              <LessonCard
                key={lesson.id}
                imageUrl={lesson.thumbnailUrl || lesson.imageUrl}
                lessonType={lesson.type}
                name={lesson.name}
                navigationPath={`/courses/${courseId}/lessons/${lesson.id}`}
                progress={isProgress ? lesson.progress : undefined}
                step={index + 1}
                surveyPerformed={isSurveyPerformed}
              />
            );
          }

          if (lesson.type === 'career_review_survey') {
            return null;
          }

          return (
            <LessonCard
              key={lesson.id}
              imageUrl={lesson.thumbnailUrl || lesson.imageUrl}
              lessonType={lesson.type}
              name={lesson.name}
              navigationPath={`/shared/student/courses/${shareId}/lessons/${lesson.id}?code=${code}`}
              showProgress={false}
              step={index + 1}
              surveyPerformed={false}
            />
          );
        })}
      </div>
    </section>
  );
};
