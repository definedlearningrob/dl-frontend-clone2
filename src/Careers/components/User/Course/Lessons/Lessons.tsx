import { useTranslation } from 'react-i18next';
import { useHistory } from 'react-router-dom';

import { LessonCard } from '@dc/shared/LessonCard';
import { TCourseLesson } from '@dc/graphql/user/queries/course';

type Props = {
  lessons: TCourseLesson[];
};

export function UserCourseLessons({ lessons }: Props) {
  const { t } = useTranslation();
  const history = useHistory();
  const isAdminApp = history.location.pathname.includes('admin');

  return (
    <section className='course-lessons'>
      <h3 className='course-lessons__heading'>{t('course.lessons.heading')}</h3>
      <div className='course-lessons__lessons'>
        {lessons
          .slice()
          .sort((a, b) => a.step - b.step)
          .map((lesson, index) => {
            const navigationPath = `${isAdminApp ? '/admin' : ''}/lessons/${lesson.id}`;

            return (
              <LessonCard
                key={lesson.id}
                imageUrl={lesson.thumbnailUrl || lesson.imageUrl}
                lessonType={lesson.type}
                name={lesson.name}
                navigationPath={navigationPath}
                showProgress={false}
                step={index + 1}
                surveyPerformed={false}
              />
            );
          })}
      </div>
    </section>
  );
}
