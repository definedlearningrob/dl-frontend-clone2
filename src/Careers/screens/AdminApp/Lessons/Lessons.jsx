import { useState, useMemo } from 'react';
import { useTranslation } from 'react-i18next';

import AdminFilters from '@dc/components/layout/Admin/Filters/Filters';
import AdminLessonsList from '@dc/components/Admin/Lessons/List/List';
import lessonsQuery from '@dc/graphql/user/queries/lessons';
import SharedFilterProvider from '@dc/shared/FilterProvider/FilterProvider';
import SharedMainContent from '@dc/shared/MainContent/MainContent';
import { FormProvider } from '@dc/hooks/useForm';
import { getLessonLabel } from '@dc/utils/lessons';
import { LESSON_TYPES } from '@dc/resources/constants';
import { ARCHIVABLE_STATUSES } from '@dc/resources/constants';

import SharedPaginatedLoader from '@shared/components/PaginatedLoader/PaginatedLoader';
import useQueryParams from '@shared/hooks/useQueryParams';
import { Select } from '@shared/components/Select';

const getLessonTypeOptions = (t) => {
  const allLessonsType = { value: null, label: t('common.fields.common.all') };
  const lessonTypes = Object.keys(LESSON_TYPES).map((key) => ({
    value: key.toLowerCase(),
    label: getLessonLabel(t, { type: key }),
  }));

  return [allLessonsType, ...lessonTypes];
};

function Lessons() {
  const { t } = useTranslation();
  const { params, updateQueryParams } = useQueryParams();
  const initialScope = params.scope
    ? ARCHIVABLE_STATUSES[params.scope]
    : ARCHIVABLE_STATUSES.ACTIVE;
  const initialLessonType = params.type
    ? ARCHIVABLE_STATUSES[params.type]
    : getLessonTypeOptions(t)[0];
  const [scope, setScope] = useState(initialScope);

  const selectScope = (selectPage) => (scope) => {
    selectPage(1);
    setScope(scope);
    updateQueryParams({ scope: scope.value });
  };

  return (
    <FormProvider>
      <SharedFilterProvider>
        {({ filter, ...filterProps }) => {
          const variables = useMemo(() => ({ filter, scope: scope.value }), [filter, scope]);

          return (
            <SharedMainContent>
              <SharedPaginatedLoader
                options={{
                  fetchPolicy: 'network-only',
                  variables,
                }}
                query={lessonsQuery}>
                {({ refetchQuery, ...pagingProps }) => (
                  <>
                    <AdminFilters>
                      <SharedFilterProvider.Search
                        field='name'
                        placeholder={t('common.placeholders.searchBy', {
                          field: t('common.fields.common.name').toLowerCase(),
                        })}
                        {...filterProps}
                      />
                      <Select
                        options={Object.values(ARCHIVABLE_STATUSES)}
                        value={scope}
                        onChange={selectScope(pagingProps.selectPage)}
                      />
                      <SharedFilterProvider.Select
                        field='type'
                        initialValue={initialLessonType}
                        options={getLessonTypeOptions(t)}
                        {...filterProps}
                      />
                    </AdminFilters>
                    <AdminLessonsList pagingProps={pagingProps} refetchQuery={refetchQuery} />
                  </>
                )}
              </SharedPaginatedLoader>
            </SharedMainContent>
          );
        }}
      </SharedFilterProvider>
    </FormProvider>
  );
}

export default Lessons;
