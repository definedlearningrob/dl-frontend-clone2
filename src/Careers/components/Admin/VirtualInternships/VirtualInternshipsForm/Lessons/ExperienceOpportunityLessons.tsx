import { useTranslation } from 'react-i18next';

import FilterProvider from '@dc/components/Admin/Lessons/Form/Items/Shared/FilterProvider/FilterProvider';
import { ARCHIVABLE_STATUSES, LESSON_TYPES } from '@dc/resources/constants';
import Lessons from '@dc/components/Admin/Courses/Form/Lessons/Lessons';
import lessonsQuery from '@dc/graphql/user/queries/lessons';

import SharedPaginatedLoader from '@shared/components/PaginatedLoader/PaginatedLoader';

import { getLessonFilters } from '../../helpers';

export const ExperienceOpportunityLessons = () => {
  const { t } = useTranslation();
  const defaultExperienceLessonType = {
    value: 'experience_opportunity',
    label: LESSON_TYPES.EXPERIENCE_OPPORTUNITY,
  };

  return (
    <FilterProvider omitUrl={true}>
      {({ SearchBar, clearFilter, filter }) => (
        <SharedPaginatedLoader
          omitUrl={true}
          options={{
            variables: {
              filter: getLessonFilters(filter, defaultExperienceLessonType),
              scope: ARCHIVABLE_STATUSES.ACTIVE.value,
            },
          }}
          query={lessonsQuery}>
          {(props) => (
            <Lessons
              SearchBar={SearchBar}
              clearFilter={clearFilter}
              fieldName='experienceOpportunityLessons'
              label={t('admin.virtualInternship.labels.experienceOpportunityLessons')}
              pagingProps={props}
              selectedLessonType={defaultExperienceLessonType}
            />
          )}
        </SharedPaginatedLoader>
      )}
    </FilterProvider>
  );
};
