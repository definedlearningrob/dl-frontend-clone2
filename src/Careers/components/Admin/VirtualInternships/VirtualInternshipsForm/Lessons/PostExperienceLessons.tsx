import { useState } from 'react';
import { useTranslation } from 'react-i18next';

import FilterProvider from '@dc/components/Admin/Lessons/Form/Items/Shared/FilterProvider/FilterProvider';
import { ARCHIVABLE_STATUSES } from '@dc/resources/constants';
import Lessons from '@dc/components/Admin/Courses/Form/Lessons/Lessons';
import lessonsQuery from '@dc/graphql/user/queries/lessons';

import SharedPaginatedLoader from '@shared/components/PaginatedLoader/PaginatedLoader';

import { getLessonFilters, getLessonTypeOptions } from '../../helpers';

export const PostExperienceLessons = () => {
  const { t } = useTranslation();
  const [selectedLessonType, setSelectedLessonType] = useState(getLessonTypeOptions()[0]);

  const selectLessonType =
    (selectPage: (page: number) => void) => (type: { value: string; label: string }) => {
      selectPage(1);
      setSelectedLessonType(type);
    };

  return (
    <FilterProvider omitUrl={true}>
      {({ SearchBar, clearFilter, filter }) => (
        <SharedPaginatedLoader
          omitUrl={true}
          options={{
            variables: {
              filter: getLessonFilters(filter, selectedLessonType),
              scope: ARCHIVABLE_STATUSES.ACTIVE.value,
            },
          }}
          query={lessonsQuery}>
          {(props) => (
            <Lessons
              SearchBar={SearchBar}
              clearFilter={clearFilter}
              fieldName='postExperienceLessons'
              label={t('admin.virtualInternship.labels.postExperienceLessons')}
              lessonTypeOptions={getLessonTypeOptions()}
              pagingProps={props}
              selectLessonType={selectLessonType}
              selectedLessonType={selectedLessonType}
            />
          )}
        </SharedPaginatedLoader>
      )}
    </FilterProvider>
  );
};
