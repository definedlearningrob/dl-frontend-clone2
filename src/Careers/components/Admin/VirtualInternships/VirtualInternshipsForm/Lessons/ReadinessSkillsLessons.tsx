import { useTranslation } from 'react-i18next';

import FilterProvider from '@dc/components/Admin/Lessons/Form/Items/Shared/FilterProvider/FilterProvider';
import { ARCHIVABLE_STATUSES, LESSON_TYPES } from '@dc/resources/constants';
import Lessons from '@dc/components/Admin/Courses/Form/Lessons/Lessons';
import lessonsQuery from '@dc/graphql/user/queries/lessons';

import SharedPaginatedLoader from '@shared/components/PaginatedLoader/PaginatedLoader';

import { getLessonFilters } from '../../helpers';

export const ReadinessSkillsLessons = () => {
  const { t } = useTranslation();
  const defaultReadinessLessonType = {
    value: 'career_readiness',
    label: LESSON_TYPES.CAREER_READINESS,
  };

  return (
    <FilterProvider omitUrl={true}>
      {({ SearchBar, clearFilter, filter }) => (
        <SharedPaginatedLoader
          omitUrl={true}
          options={{
            variables: {
              filter: getLessonFilters(filter, defaultReadinessLessonType),
              scope: ARCHIVABLE_STATUSES.ACTIVE.value,
            },
          }}
          query={lessonsQuery}>
          {(props) => (
            <Lessons
              SearchBar={SearchBar}
              clearFilter={clearFilter}
              fieldName='readinessSkillsLessons'
              label={t('admin.virtualInternship.labels.readinessSkillsLessons')}
              pagingProps={props}
              selectedLessonType={defaultReadinessLessonType}
            />
          )}
        </SharedPaginatedLoader>
      )}
    </FilterProvider>
  );
};
