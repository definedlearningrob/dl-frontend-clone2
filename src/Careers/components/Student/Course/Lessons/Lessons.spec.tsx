import { screen } from '@testing-library/react';
import React from 'react';
import { Route } from 'react-router-dom';

import { StudentCourseLessons } from '@dc/components/Student/Course/Lessons/Lessons';
import { renderWithRouterAndReduxProvider } from '@dc/utils/test';

const lessons = [
  {
    careerReviewSurvey: {
      performed: false,
      version: 1,
    },
    progress: {
      total: 2,
      submitted: 1,
    },
    id: '2',
    imageUrl: 'image-url',
    name: 'Lesson name 2',
    step: 2,
    type: 'lesson-type',
    thumbnailUrl: 'thumbnail-url',
  },
  {
    careerReviewSurvey: {
      performed: false,
      version: 1,
    },
    progress: {
      total: 2,
      submitted: 1,
    },
    id: '1',
    imageUrl: 'image-url',
    name: 'Lesson name',
    step: 1,
    type: 'lesson-type',
    thumbnailUrl: 'thumbnail-url',
  },
];

describe('Lessons', () => {
  it('should render a list of lessons correctly', () => {
    const { container } = renderWithRouterAndReduxProvider(
      <Route path='course/:id'>
        <StudentCourseLessons courseProgress={{ total: 4, submitted: 2 }} lessons={lessons} />
      </Route>,
      { route: 'course/123' }
    );

    expect(screen.getAllByRole('link')).toHaveLength(2);

    expect(container).toMatchSnapshot();
  });
});
