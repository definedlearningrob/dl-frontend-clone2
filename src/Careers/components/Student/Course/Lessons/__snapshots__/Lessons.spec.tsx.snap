// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lessons should render a list of lessons correctly 1`] = `
<div>
  <section
    class="course-lessons"
  >
    <h3>
      Lessons
    </h3>
    <div
      class="course-lessons__lessons"
    >
      <a
        aria-label="Lesson name"
        class="card"
        href="/courses/123/lessons/1"
      >
        <div
          class="step started"
        >
          1
        </div>
        <div
          class="imageWrapper"
        >
          <img
            alt="Lesson name"
            class="image"
            src="thumbnail-url"
          />
        </div>
        <section
          class="flex flex-col justify-between flex-1 pt-xs px-xs pb-xxs min-h-[100px]"
        >
          <h5
            class="name"
          >
            Lesson name
          </h5>
          <p
            class="progressValue"
          >
            50%
          </p>
        </section>
        <section
          class="relative bg-primary-200 rounded-b-sm text-primary-500 before:h-xxxs before:bg-neutral-400 before:inset-0 before:absolute bg-secondary-200 text-secondary-500"
        >
          <div
            class="progressBar started"
            style="width: 50%;"
          />
          <div
            class="flex items-center text-xxs p-xs xxxl:text-xs xxxl:py-sm xxxl:px-xs leading-lg"
          >
            Continue
            <div
              class="icon footerIcon -xs-size"
              data-testid="icon"
            >
              <div />
            </div>
          </div>
        </section>
      </a>
      <a
        aria-label="Lesson name 2"
        class="card"
        href="/courses/123/lessons/2"
      >
        <div
          class="step started"
        >
          2
        </div>
        <div
          class="imageWrapper"
        >
          <img
            alt="Lesson name 2"
            class="image"
            src="thumbnail-url"
          />
        </div>
        <section
          class="flex flex-col justify-between flex-1 pt-xs px-xs pb-xxs min-h-[100px]"
        >
          <h5
            class="name"
          >
            Lesson name 2
          </h5>
          <p
            class="progressValue"
          >
            50%
          </p>
        </section>
        <section
          class="relative bg-primary-200 rounded-b-sm text-primary-500 before:h-xxxs before:bg-neutral-400 before:inset-0 before:absolute bg-secondary-200 text-secondary-500"
        >
          <div
            class="progressBar started"
            style="width: 50%;"
          />
          <div
            class="flex items-center text-xxs p-xs xxxl:text-xs xxxl:py-sm xxxl:px-xs leading-lg"
          >
            Continue
            <div
              class="icon footerIcon -xs-size"
              data-testid="icon"
            >
              <div />
            </div>
          </div>
        </section>
      </a>
    </div>
  </section>
</div>
`;
